Support Standard Name Modifers (not units!)

lbdreyer opened the pull request:
CF conventions support a set of standard_name modifiers.  See [3.3. Standard Name](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#standard-name) for more details, specifically:

> A standard name is associated with a variable via the attribute standard_name which takes a string value comprised of a standard name optionally followed by one or more blanks and a standard name modifier (a string value from Appendix C, Standard Name Modifiers).

This PR modifies the CFVariableMixin standard_name setter (and the pyke rules) to support standard name modifiers. 

I found there wasn't much testing of cube/coord names so I have split this PR into two commits:
1. The first commit has no functional changes. It just adds testing of the pyke rules to show that they don't break later when I add the functional changes
2. The second commit adds the changes to include the standard_name modifiers

Note that standard name modifiers have associated units (See [CF Appendix C](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#standard-name-modifiers)) however I have not included that in this PR as I am only modifying the standard_name setter which at the moment doesn't have any requirement on the unit

lbdreyer commented:
Ping @bjlittle !

bjlittle commented:
@lbdreyer Okay, I think that I'm pretty much there...

As it happens, there wasn't any unit test coverage of `CellMethods` (there is for parsing, but not construction and usage), but I've added the minimal test coverage for the changes in this PR.

Being a good citizen, I'll add test coverage in a separate PR.

I've also not updated the docs either, just so that this PR can get banked sooner rather than later, but happy to make a follow-up PR, if you think that's appropriate... :smile:

bjlittle commented on lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb:
```diff
@@ -1523,20 +1527,25 @@ fc_extras
+                            attributes['invalid_standard_name'] = standard_name
+                            standard_name = coord_name
+                        else:
+                            standard_name = None
```
The previous behaviour also set the `long_name = standard_name` here i.e.,
```python
else:
    standard_name = None
    long_name = standard_name
```
This is still relevant, right?

lbdreyer replied:
I removed it as I don't see the point in that line.
Just looking at the old code (you have to scroll the below snippet to see the last line):
https://github.com/SciTools/iris/blob/ec46449b9705789268b77089a2161ff2c1d605c5/lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb#L1527-L1539
One line 1527 it says
`if long_name is not None`
so line 1539 is within the `else` of that statement, implying that `long_name=None`. Which means that line 1539  was setting long_name to be None, but it already was None...

bjlittle replied:
Yup, agreed :+1:

bjlittle commented on lib/iris/tests/integration/test_netcdf.py:
```diff
@@ -429,5 +429,15 @@ def test_scalar_cube_save_load(self):
+        with self.temp_filename(suffix='.nc') as fout:
+            iris.save(cube, fout)
+            detection_limit_cube = iris.load_cube(fout)
+            self.assertEqual(detection_limit_cube.standard_name, standard_name)
```
Nice :+1:

bjlittle commented on lib/iris/tests/unit/cube_coord_common/test_CFVariableMixin.py:
```diff
@@ -0,0 +1,48 @@
+class Test_standard_name__setter(tests.IrisTest):
+    def test_valid_standard_name(self):
+        cf_var = CFVariableMixin()
+        cf_var.standard_name = 'air_temperature'
```
Shouldn't you then `self.assertEqual(cf_var.standard_name, 'air_temperature')` to ensure that we get back what we set?

lbdreyer replied:
I guess my concern with that was that we are then testing the standard_name getter and setter at the same time, which doesn't feel very unit-testy but I guess we should test not just that setting the standard doesn't raise errors, but also that it is done correctly and it's hard to do that in isolation from testing the getter.

bjlittle commented on lib/iris/tests/unit/cube_coord_common/test_CFVariableMixin.py:
```diff
@@ -0,0 +1,48 @@
+
+    def test_none_standard_name(self):
+        cf_var = CFVariableMixin()
+        cf_var.standard_name = None
```
Also, include a `self.assertIsNone(cf_var.standard_name)`?

bjlittle commented on lib/iris/tests/unit/fileformats/pyke_rules/compiled_krb/fc_rules_cf_fc/test_build_cube_metadata.py:
```diff
@@ -35,33 +35,33 @@
-        engine = mock.Mock(
-            cube=Cube([23]),
-            cf_var=cf_var)
+    return engine
```
Wondered why you pulled this out of `TestInvalidGlobalAttributes`, but you're also using it in the new test class `TestCubeName`... nice :+1:

bjlittle commented:
@lbdreyer What's the strategy for checking the `units` of a `standard_name` with a standard name modifier?

Separate PR?

bjlittle commented on lib/iris/tests/unit/fileformats/pyke_rules/compiled_krb/fc_rules_cf_fc/test_get_names.py:
```diff
@@ -0,0 +1,258 @@
+        inp = ('latitude_coord', 'lat_long_name', 'lat_var_name', 'latitude')
+        exp = ('latitude', 'lat_long_name', 'lat_var_name',
+               {'invalid_standard_name': 'latitude_coord'})
+        self.check_names(inp, exp)
```
@lbdreyer Whoa... nice test coverage :tada:

lbdreyer commented on lib/iris/_cube_coord_common.py:
```diff
@@ -28,6 +28,27 @@
+def get_valid_standard_name(name):
+    # Standard names are optionally followed by a standard name
+    # modifier, separated by one or more blank spaces
+    name_parts = name.split(' ')
```
If `name` has leading or trailing white space it should cause issues, I think, as then the name is not valid.
```
bad_name = ' air_temperature'
>>> bad_name.split(' ')
['', 'air_temperature']
```
should raise an error, as should 
```
>>> bad_name = 'air_temperature '
>>> bad_name.split(' ')
['air_temperature', '']
```

lbdreyer commented on lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb:
```diff
@@ -1186,13 +1187,16 @@ fc_extras
+                    if cube.long_name is not None:
+                        cube.attributes['invalid_standard_name'] = standard_name
+                    else:
+                        cube.long_name = standard_name
```
No? The diff that is being displayed is makign this a bit hard to read, but basically what was previously lines 1527-1538 
https://github.com/SciTools/iris/blob/ec46449b9705789268b77089a2161ff2c1d605c5/lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb#L1527-L1539
https://github.com/SciTools/iris/blob/b6775634eccbe29a45519c85a0da99a36dbc4886/lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb#L1536-L1547
But they have been indented out once more because they are in the else of the except block.

One change I have made is I have removed line 1539 in the old code (see the last line in the first snippet above) because this line seemed pointless? It was setting long_name to be None even though it already was None?

lbdreyer replied:
I don't believe so?
What were lines 1192-1195 in the old code
https://github.com/SciTools/iris/blob/ec46449b9705789268b77089a2161ff2c1d605c5/lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb#L1192-L1195
Are now lines 1196-1199 in the new code
https://github.com/SciTools/iris/blob/b6775634eccbe29a45519c85a0da99a36dbc4886/lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb#L1196-L1199

so no change, just an extra indentation due to being in the `else` of the `if`statement in the `except` part

lbdreyer commented on lib/iris/_cube_coord_common.py:
```diff
@@ -28,6 +28,27 @@
+                    standard_name_modifier not in
+                    iris.fileformats.cf.STD_NAME_MODIFIERS):
+                raise ValueError(error_msg)
+        return name
```
Yeah I agree, it would be nicer.I just couldn't quite think through how to do that, but I will give it another look!

bjlittle replied:
@lbdreyer I think that all you need to do is simply move the `return name` outside the `if ... else ...`, as your not changing the contents of `name`, the output of the function is just the same as the input.

If the input `std_name` (from `name`) is valid, then it will pass through the `if ...`, and then if the `standard_name_modifier` is valid it will simply drop through and `return name`. On the otherhand, if the modifier is not valid an exception is raised and the `return name` is never be reached.

If the input `name` is invalid, then it will pass through to the `else ...` where an exception is raised and the `return name` will never be reached.

Should just work, right? Check my logic is right, though...

lbdreyer commented:
> @lbdreyer What's the strategy for checking the `units` of a `standard_name` with a standard name modifier?
> 
> Separate PR?

I have raised this in a separate issue: https://github.com/SciTools/iris/issues/3408

lbdreyer commented on lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb:
```diff
@@ -1186,13 +1187,16 @@ fc_extras
+                cube.standard_name = get_valid_standard_name(standard_name)
+            except ValueError as e:
+                if str(e) != '{!r} is not a valid standard_name'.format(standard_name):
+                    raise
```
My concern is that with the `try ... except`_any_ ValueError would be ignored. However, I suppose that any issues with setting the standard_name, e.g an issue with the regexp matching (in the upcoming commit "review actions") should be skipped and go on to try to set the long name instead

lbdreyer commented:
@bjlittle I believe I have addressed your review comments in the third commit ["review actions"](https://github.com/SciTools/iris/pull/3393/commits/101dc5cb2f2a9a35721eb119757ddc623ab167dd)

Note: I had to rebase the first two commits due a conflict that was caused when #3399 was merged, but those commits should be the same as they were before.

bjlittle commented:
@lbdreyer Awesome, thanks! :tada:
